scripts/common/timer.ts

```diff
@@ -88,43 +96,321 @@ export interface RunSubsegment {
+	// e.g. how fast did they get their 2nd collectible, 3rd collectible, etc...
+	return [baseSeg.subsegments[baseSubsegIdx], compSeg.subsegments[baseSubsegIdx] ?? null];
 }
 
-export function getSegmentName(segmentIndex: number, subsegmentIndex: number): string {
-	return subsegmentIndex >= 1 ? `${segmentIndex + 1}-${subsegmentIndex}` : segmentIndex.toString();
+/**
+ * Try find the index of a subsegment hit previously in the given splits, taking unordered and missing checkpoints into
+ * account, returning last subsegment index in the previous segment if a starting subsegment is provided.
+ * @returns Tuple of [majorNum, minorNum]
+ */
+export function findPreviousSubsegment(splits: RunSplits, majorNum: number, minorNum: number): [number, number] {
+	if (minorNum === 1) {
+		if (majorNum === 1) {
+			// If we're at the very start of the run, there is no previous subsegment
+			return [0, 0];
+		}
+
+		// If minorNum == 1 we're in a starting segment, so the previous subsegment is always the last subsegment in the
+		// last segment's subsegments. Segments are never optional, and must have had a start segment so both index
+		// accesses must be valid.
+		return [majorNum - 1, splits.segments[majorNum - 2].subsegments.length];
+	}
+
+	const segment = splits.segments[majorNum - 1];
+
+	// Find last subsegment in subsegments, i.e. in chronological order -- for unordered/missing subsegments this
+	// could be < minorNum - 1
+	const subsegment = segment.subsegments[segment.subsegments.findIndex((ss) => ss.minorNum === minorNum) - 1];
+
+	// Weird if this happens, our JS is probably bad!
+	if (!subsegment) {
+		throw new Error(`Missing subsegment for majorNum: ${majorNum}, minorNum: ${minorNum}`);
+	}
+
+	return [majorNum, subsegment.minorNum];
 }
 
-export interface Comparison {
-	baseRun: RunMetadata;
-	comparisonRun: RunMetadata;
-	diff: number;
-	overallSplit: ComparisonSplit;
-	segmentSplits: ComparisonSplit[][];
+export function getSplitName(
+	splits: RunSplits,
+	majorNum: number,
+	minorNum: number,
+	segmentsCount: number,
+	segmentsCheckpointsCount: number
+): string {
+	if (majorNum === 1 && minorNum === 1) return '';
+
+	// Single segment run (e.g. linear surf map): previous minorNum
+	// Note that splits.segments.length is viable here since only contains splits hit so far.
+	if (segmentsCount === 1) {
+		const [, prevMin] = findPreviousSubsegment(splits, majorNum, minorNum);
+		return prevMin.toString();
+	}
+
+	// Multiple segments, multiple subsegments
+	// (e.g. RJ map with courses): "X-Y" for previous majorNum, previous minorNum
+	if (segmentsCheckpointsCount > 1) {
+		// For these zones, we want:
+		//  S1----CP1----CP2----S2----CP1----CP2----END
+		// (n/a)  1-1    1-2    1-3   2-1    2-2    2-3
+		// So:
+		// Maj 1, Min 1 => N/A      Maj 1, Min 2 => 1-1      Maj 1, Min 3 => 1-2
+		// Maj 2, Min 1 => 1-3      Maj 2, Min 2 => 2-1      Maj 2, Min 3 => 2-2
+		// END => 2-3
```

Comment:
Maybe the end zone splits should display "2-End" instead of "2-3"? Idk, I don't have a strong preference either way.

Reply:
This is used for the name of the checkpoint you're current on, or was just completed.

If you do a surf map with two stages, no minor CPs, when you complete stage 1, the comparison will be named "1", when you complete stage 2 (and thus complete the map), it shows "2". Showing "End" instead of "2" seems very weird in that case.

Similarly if each stage had an extra minor CP, the comparisons would show "1-1", "1-2", "2-1", "2-2".